Refine CONTRIBUTING.md to include a clear section for issues and pull requests. Add a code quality section (#140)
FuelLabs/sway-applications: CONTRIBUTING.md

### Code Structure

In order to structure your code there is only 1 rule that you must remember and that is that everything must always be ordered alphabetically - as long as there isn't a dependency error or a better name for a default field of an enum. The reason is that it provides a structure that almost everyone will be familiar with. This means that nobody has to learn how you have decided to structure your code and thus where to look for a variable, function, import etc. This makes it really easy to navigate your entire project and in turn navigate all projects without having to memorize the intricacies of each project.

To be explicit

- All dependencies must be declared in alphabetical order just below the declaration of the program type e.g. below `contract;`
- All imports must be in alphabetical order and any modules within those imports must be in alphabetical order
- All functions, structs, enums, fields inside structs and enums, parameters of a function, parameters in documentation etc. must be in alphabetical order in all files (this does not apply to variables inside a function)

Example:

> ```rust

Reviewer: I think it would be smart to have this in an external file and included in CI, much like how examples are written in the sway book

Author: :) Probably, I did write this off-hand without checking if it even compiles

Reviewer 2: Unfortunately, plain markdown doesn't support that kind of include. We could consider making a Sway Applications Book though!

Author: This warrants a separate issue. Will be done in #141

storage {
   games_played: u64,
   player_two_wins: u64,
   player_one_wins: u64,
   ...
}

impl ConnectFour for Contract {

   fn create_game(player_two: Player, player_one: Player) -> Game {
      // Perform a check on each player address to see if they are blacklisted

      // owl
   }

   fn move(column: u64, game: Game) -> Game {
      // Perform a check to see if the game has ended
      // Perform a check to see if the position is valid
      
      // owl
   }

   // rest of owl

}
```

The only other aspect to remember is to use the formatter before commiting your work.

- `cargo fmt` to format your `Rust` files (SDK tests)
- `forc fmt` to format your `Sway` files

### Documentation

#### Code

For guidance refer to how [Rust](https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html) documents code.

There are two ways to document your code

- Using `///` for official documentation
  - This allows tools to be used and parse out your documentation automatically
- Using `//` for comments
  - These are helper comments within the code to guide the developer who is reading your code
  - They do not get parsed out via automatic doc generation

> Each function that is written should be documented (the `///`). The only exception applies to the functions inside the `impl` of your contract (notice that the example above does not have an documentation, that would be on the `abi` that is imported). The documentation for those functions should be on the `abi` because a developer will have access to your `abi` and not necessarily your implementation.

Reviewer: ```suggestion
Each function that is written should be documented (the `///`). The only exception applies to the functions inside the `impl` of your contract (notice that the example above does not have any documentation, that would be on the `abi` that is imported). The documentation for those functions should be on the `abi` because a developer will have access to your `abi` and not necessarily your implementation.
```

> When writing documentation make sure that your arguments are in alphabetical order and any assertion that you have in your function is also documented in the same order. It makes it easier to read the code when you see everything in sequential order rather than searching for where an assertion is placed in the function.

Reviewer: ```suggestion
When writing documentation make sure that your arguments are in alphabetical order and any assertions that you have in your function are also documented in the same order. It makes it easier to read the code when you see everything in sequential order rather than searching for where an assertion is placed in the function.
```

   ///
   /// # Arguments
   ///
   /// - `column` - The column to place a marker in, range 0 <= column < 8
   /// - `game` - The game to make a move in
   ///
   /// # Reverts
   ///
   /// - When a game has ended in a player winning or a draw
   /// - When a marker is placed into a `column` that is full
   fn move(column: u64, game: Game) -> Game;
}
```

In addition to documenting your functions make sure to document your data structures, events, errors etc. if needed. It is your job to explain to the reader what your code is an what it does so do not make the reader guess. It might be obvious to you but not to the reader.

That being said, there is good documentation and bad documentation.

Example

```rust
// This is bad. It's just repeating what I can already read from the field names
pub struct Item {
   /// Identifier
   id: u64,

   /// Quantity
   quantity: u64,
}

// This is better. It tells me the context of what the fields are
pub struct Item {
   /// Unique identifier used to retrieve the item from a vector of items held in storage
   id: u64,

   /// The number of remaining items left in production
   quantity: u64,
}
```

> Documenting your code does not stop at writing documentation. It is also important to name your variables, functions and data structures appropriately. This is a difficult task and it can be argued to be a skill and an art because if your name is too verbose then it will be difficult to read (because of how much screenspace it takes up) and annoying to use while an abbreviated name requires insider knowledge to be able to infer what the variable is (you could document it extensively however that still means the reader must remember what the documentation says).

Reviewer: ```suggestion
Documenting your code does not stop at writing documentation. It is also important to name your variables, functions and data structures appropriately. This is a difficult task and it can be argued to be a skill and an art. If your name is too verbose, then it will be difficult to read (because of how much screenspace it takes up) and annoying to use. On the other hand, an abbreviated name requires insider knowledge to be able to infer what the variable is (you could document it extensively however that still means the reader must remember what the documentation says).
```

> There are two general rules to follow

Reviewer: ```suggestion
There are two general rules to follow:
```
